ludwig/train.py

```diff
@@ -396,6 +400,68 @@ def full_train(
     )
 
 
+def kfold_cross_validate(
+                model_definition,
+                model_definition_file=None,
+                data_csv=None,
+                data_train_csv=None,
+                data_validation_csv=None,
+                output_directory='results',
+                k_fold=None,
+                **kwargs
+):
+    logger = logging.getLogger('ludwig')
+    logger.info('starting {:d}-fold cross validation'.format(k_fold))
+
+    # create output_directory if not available
+    if not os.path.isdir(output_directory):
+        os.mkdir(output_directory)
+
+    # read in data to split for the folds
+    data_df = pd.read_csv(data_train_csv)
+
+    # place each fold in a separate directory
+    data_dir = os.path.dirname(data_train_csv)
+    kfold_training_stats = {}
+    for train_index, test_index, fold_num in generate_kfold_splits(data_train_csv, k_fold):
+        with tempfile.TemporaryDirectory(dir=data_dir) as temp_dir_name:
+            # save training and validation subset for the fold into a temporary directory
+            train_csv_fp = os.path.join(temp_dir_name, 'train_fold.csv')
+            test_csv_fp = os.path.join(temp_dir_name, 'test_fold.csv')
+            logger.info("\n\n>>>>> for fold {:d} created temporary directory: {}".format(fold_num, temp_dir_name))
+            data_df.iloc[train_index].to_csv(train_csv_fp, index=False)
+            data_df.iloc[test_index].to_csv(test_csv_fp, index=False)
+
+            # train and validate model on this fold
+            logger.info("training on fold {:d}".format(fold_num))
+            (model,
+             preprocessed_data,
+             _,
+             train_stats,
+             model_definition) = full_train({},
+                                            model_definition_file=model_definition_file,
+                                            data_train_csv = train_csv_fp,
+                                            data_test_csv = test_csv_fp,
+                                            experiment_name='cross_validation',
+                                            model_name='fold_' + str(fold_num),
+                                            output_directory=os.path.join(temp_dir_name,'results'))
+
+            # score on hold out fold
+            preds = model.predict(preprocessed_data[2],
+                                  model_definition['training']['batch_size'])
+
+            # augment the training statistics with scoring metric fron the hold out fold
+            train_stats['fold_metric'] = preds['combined']
+
+            # collect training statistics for this fold
+            kfold_training_stats['fold_'+str(fold_num)] = train_stats
+
```

Comment:
before saving i would add an additional `'overall'` key to the kfold stats that for each output feature takes the overall and adds mean and std.
Something like:
```
overall_kfold_stats = {}
for fold_stats in kfold_training_stats:
    for of in fold_stats:
        if OVERALL in of:
            for metric in of[OVERALL]:
                if metric not in overall_kfold_stats:
                    overall_kfold_stats[metric] = []
                overall_kfold_stats[metric].append(of[OVERALL][metric])
for metric in overall_kfold_stats:
    mean = np.mean(overall_kfold_stats[metric])
    std = np.std(overall_kfold_stats[metric])
    del overall_kfold_stats[metric]
    overall_kfold_stats[metric + "_mean"] = mean
    overall_kfold_stats[metric + "_std"] = std
kfold_training_stats[OVERALL] = overall_kfold_stats
```

Reply:
Added `'overall'` key.  This contains the mean and standard deviation for the various metrics (loss, accuracy) from the hold out fold.  Here is an example:
```
   "overall": {
        "Survived": {
            "accuracy_mean": 0.39955106621773284,
            "accuracy_std": 0.015632310075403076,
            "loss_mean": 6.96722408684282,
            "loss_std": 0.5409600467679053
        },
        "combined": {
            "accuracy_mean": 0.39955106621773284,
            "accuracy_std": 0.015632310075403076,
            "loss_mean": 6.96722408684282,
            "loss_std": 0.5409600467679053
        }
    }

```
The structure of the `fold_training_statistics.json` file is now
```
{
  "fold_1": {...}
 "fold_2": {...}
 "fold_3": {...}
 "overall": {...}
}
```

Reply:
looks good